src/uu/numfmt/src/numfmt.rs

```diff
@@ -43,19 +54,67 @@ where
     R: BufRead,
 {
     let mut lines = input.lines();
-    for (idx, line) in lines.by_ref().enumerate() {
+
+    for line in lines.by_ref().take(options.header) {
         match line {
-            Ok(l) if idx < options.header => {
-                println!("{}", l);
-                Ok(())
+            Ok(l) => println!("{}", l),
+            Err(e) => return Err(Box::new(NumfmtError::IoError(e.to_string()))),
+        };
+    }
+
+    let mut has_failed_flag = false;
+
+    for line_result in lines.by_ref() {
+        match line_result {
+            Ok(line) => {
+                let format_result = format_and_handle_validation(&line, options);
+
+                if format_result.is_err() {
+                    match options.invalid {
+                        InvalidModes::Abort => {
+                            return format_result;
+                        }
+                        InvalidModes::Fail => has_failed_flag = true,
+                        _ => {}
+                    }
+                }
             }
-            Ok(l) => match format_and_print(&l, options) {
-                Ok(_) => Ok(()),
-                Err(e) => Err(NumfmtError::FormattingError(e.to_string())),
-            },
-            Err(e) => Err(NumfmtError::IoError(e.to_string())),
-        }?;
+            Err(e) => {
+                return Err(Box::new(NumfmtError::IoError(e.to_string())));
+            }
+        }
+    }
+
+    if has_failed_flag {
+        return Err(Box::new(NumfmtError::FailModeError()));
+    }
+
+    Ok(())
+}
+
+fn format_and_handle_validation(input_line: &str, options: &NumfmtOptions) -> UResult<()> {
+    let handled_line = format_and_print(input_line, options);
+
+    if let Err(error_message) = handled_line {
+        match options.invalid {
+            InvalidModes::Abort => {
+                return Err(Box::new(NumfmtError::FormattingError(error_message)));
+            }
+            InvalidModes::Fail => {
+                eprintln!("numfmt: {}", error_message);
+                println!("{}", input_line);
+                return Err(Box::new(NumfmtError::FormattingError(error_message)));
+            }
+            InvalidModes::Ignore => {
+                println!("{}", input_line);
+            }
+            InvalidModes::Warn => {
+                eprintln!("numfmt: {}", error_message);
```

Comment:
This can be done with [`show_error!`](https://docs.rs/uucore/0.0.16/uucore/macro.show_error.html), which will also change the name depending on how the utility is called.